ocaml/alerts/expiry_alert.ml

```diff
@@ -0,0 +1,126 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module XenAPI = Client.Client
+module Date = Xapi_stdext_date.Date
+
+let seconds_per_day = 3600. *. 24.
+
+let days_until_expiry epoch expiry =
+  (expiry /. seconds_per_day) -. (epoch /. seconds_per_day)
+
+let all_messages rpc session_id =
+  XenAPI.Message.get_all_records ~rpc ~session_id
+
+let related_messages msg_obj_uuid related_message_name_list all_msgs =
+  List.filter
+    (fun (_ref, record) ->
+      let related name = List.mem name related_message_name_list in
+      record.API.message_obj_uuid = msg_obj_uuid
+      && related record.API.message_name
+    )
+    all_msgs
+
+let message_body msg expiry =
+  Printf.sprintf "<body><message>%s</message><date>%s</date></body>" msg
+    (Date.to_string expiry)
+
+let expired_message obj = Printf.sprintf "The %s has expired." obj
+
+let expiring_message obj = Printf.sprintf "The %s is expiring soon." obj
+
+let generate_alert now expiry expired_message_id expiring_conditions
+    alert_obj_description =
+  let remaining_days =
+    days_until_expiry (Date.to_float now) (Date.to_float expiry)
+  in
+  let sorted_alert_conditions =
+    List.sort
+      (fun (remaining_days_a, _) (remaining_days_b, _) ->
+        remaining_days_a - remaining_days_b
+      )
+      ((0, expired_message_id) :: expiring_conditions)
```

Comment:
I think this special case could be avoided from beginning. Generally I think if it could be merged with other cases eventually, it should be possible to treat them in unified way in the first place.

Reply:
With `expired_message_id` as an individual parameter avoid having to check if an expired message id is set in the list, as I think an expired message is required, while expiring messages are optional.